### ISSUE TITLE
CRI: Define CRI log format better

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#44976]
Current [cri logging format](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/kubelet-cri-logging.md) doesn't mention how to handle new line in the log.

Currently, it seems that we assume that each cri log line corresponds to a container output line. In this way, we don't need to handle newline specially.

However, there are 2 problems:
1) If the container only outputs one line without newline at the end, the log generated won't have newline, either. (cri-o hit this in their node e2e test)
2) If the log line is too long, runtime usually has to split one line into several lines.

Both cases will confuse kubelet.

It would be better to:
1) Let runtime escape all newlines in log content, and add newline to the end as part of CRI logging format. 
2) Let kubelet unespace the newlines inside the log content.

/cc @yujuhong @feiskyer @mrunalp /cc @kubernetes/sig-node-api-reviews

**Comment by mrunalp** [comment]
cc: @cyphar @runcom

**Comment by mrunalp** [comment]
We hit this because there is no newline in this test:
```
 Failure [9.337 seconds]
[k8s.io] Projected
/root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:656
  should project all components that make up the projection API [Conformance] [Volume] [Projection] [It]
  /root/gosrc/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/projected.go:1023

  Expected error:
      <*errors.errorString | 0xc4204b86d0>: {
          s: "expected \"projected-volume-9205b84d-2ac6-11e7-9966-54ee75105a2b\" in container output: Expected\n    <string>: \nto contain substring\n    <string>: projected
-volume-9205b84d-2ac6-11e7-9966-54ee75105a2b",
      }
      expected "projected-volume-9205b84d-2ac6-11e7-9966-54ee75105a2b" in container output: Expected
          <string>: 
      to contain substring
          <string>: projected-volume-9205b84d-2ac6-11e7-9966-54ee75105a2b
  not to have occurred
```

Adding `&& echo hi` passes the test for cri-o. 

Another simple option is to just not look for \n in the kubelet code for the last line in the log file.

**Comment by feiskyer** [comment]
>It would be better to:
>
>1. Let runtime escape all newlines in log content, and add newline to the end as part of CRI logging format. (cri-o hit this in their node e2e test)
>2. Let kubelet unespace the newlines inside the log content.

2 looks more reasonable to me. There is usually no newline at the end of log files.

**Comment by cyphar** [comment]
Personally I'd prefer if we didn't have this silly newline-separation logic (each newline having to be prefixed with a timestamp)_ in `cri-o` (runtimes should be able to just dump the contents of the buffers they read from a container's stdio and kubelet should be able to handle that) but there are a few complications this raises (generating JSON or similar within C is not fun especially if the runtime would have to escape strings).

However for the problems you outlined:

> If the container only outputs one line without newline at the end, the log generated won't have newline, either.

Kubelet shouldn't really be caring about the trailing newline of a file IMO. In order to avoid the continuation problem you mention, `cri-o` doesn't add newlines (unless you have interleaved output from two different streams, which can't be solved without my original comment about not having newline-separated logs).

> If the log line is too long, runtime usually has to split one line into several lines.

This is only truly necessary for the interleaved case, but I'd assume most runtimes would write a final newline to the log (it makes the implementation easier).

**Comment by Random-Liu** [comment]
> 2 looks more reasonable to me. There is usually no newline at the end of log files.

Actually, there is only one solution proposed here. Both kubelet and the runtime need to take action.

**Comment by Random-Liu** [comment]
> Personally I'd prefer if we didn't have this silly newline-separation logic (each newline having to be prefixed with a timestamp)_ in cri-o (runtimes should be able to just dump the contents of the buffers they read from a container's stdio and kubelet should be able to handle that) but there are a few complications this raises (generating JSON or similar within C is not fun especially if the runtime would have to escape strings).

Do you mean that we should let runtime just dump purely container output without any other thing? I think we still need the timestamp which is discussed in the original log proposal.

> Kubelet shouldn't really be caring about the trailing newline of a file IMO. In order to avoid the continuation problem you mention, cri-o doesn't add newlines (unless you have interleaved output from two different streams, which can't be solved without my original comment about not having newline-separated logs).

Do you mean that Kubelet should not try to read newline at the end of the file? The problem is that `kubectl log` may return incomplete last log line. This may be acceptable.

> This is only truly necessary for the interleaved case, but I'd assume most runtimes would write a final newline to the log (it makes the implementation easier).

The problem what I mentioned is that:
1) If runtime dumps stderr and stdout into the same file, a very looong line from one stream will block another;
2) If runtime dumps stderr and stdout into different files, then it introduces extra work and overhead to merge them (CRI logging format needs only one log file).

**Comment by mrunalp** [comment]
> Do you mean that Kubelet should not try to read newline at the end of the file? The problem is that kubectl log may return incomplete last log line. This may be acceptable.

:+1: to this. This is the simplest fix. It isn't necessarily incomplete as the output doesn't always have a newline.

**Comment by crassirostris** [comment]
@Random-Liu I want to make sure I understand the problem correctly, could you please verify?

If container writes `foo` to the `stdout` without a newline, runtime will print a following message to the log file: `<metadata> foo<EOF>` (no newline). If container then writes `bar` to the `stdout`, it's not clear should runtime write `<metadata> foobar<EOF>` or `<metadata> foo\n<metadata> foo<EOF>`

And if between those two messages container writes `baz` to the `stderr`, it gets even more complicated, because we have to write it as `<metadata> foo\n<metadata> baz\n<metadata> bar`

Then the problem is, it's unclear how to distinguish between `foo\nbar` and `foobar` on the reading stage. Wrapping line endings looks like a solution, but it's a breaking change to the format

**Comment by Random-Liu** [comment]
If we don't escape `\n`, no matter how long a line is, we could not break it.

On the runtime implementation side, it means that:
1) If we have 2 threads, one for stderr, one for stdout. We have to keep the whole log line in memory, and log it once to make sure that the log line will not be broken by stderr. This may cause huge memory usage if the line is too long;
2) If we have 1 thread for both stderr and stdout. We can keep logging the stdout line, and block stderr. However, if the log line is too long, this may block stderr for too long time, and eventually block the application.

This is purely extreme case, but to protect the system we may want to consider it.

**Comment by crassirostris** [comment]
@Random-Liu I think second approach is more expected. I mean consider a regular process writing both to `stdout` and `stderr`. It's going to be blocked on write to `stderr` if message being written to `stdout` is too large.

**Comment by cyphar** [comment]
@crassirostris I don't believe that's accurate, as the kernel internally buffers writes (and only writes below a certain size are guaranteed to be atomic). In particular, I would be incredibly surprised if it was that easy to block.

Personally I think the only way of handling this sanely is to have something like a JSON blob for each "line" or entry in the log, which has an escaped string for the buffer being read from the container pipes. I don't like that we would have to do this, but I don't see a way of handling the log using valid output as terminators for entries...

> On the runtime implementation side, it means that:

On the other hand, if we implement proper escaping then the runtime can just work in fixed-size chunks and not have to worry about either of the cases you mentioned.

**Comment by Random-Liu** [comment]
>  I don't believe that's accurate, as the kernel internally buffers writes (and only writes below a certain size are guaranteed to be atomic). In particular, I would be incredibly surprised if it was that easy to block.

@cyphar Yeah, linux has a 4096 PIPE_BUF limit. Even option 1 is not actually working.

> On the other hand, if we implement proper escaping then the runtime can just work in fixed-size chunks and not have to worry about either of the cases you mentioned.

Yeah, I also think we should escape the new line in the log, so that runtime could cut the log line if needed.

**Comment by crassirostris** [comment]
@cyphar What I meant is user space level behavior, [like in c++](http://en.cppreference.com/w/cpp/io/cerr) when stdout is flushed before writing to stderr and there's no concurrency on write. I agree though that it's not generic enough for all possible applications.

What I worry about is the consumption step. Interleaving messages will force the log consumer (say, logging agent exporting logs to a remote destination) to buffer and piece together streams instead, complicating logic and causing memory spikes and OOM errors (e.g. if app never writes `\n`). But considering what @Random-Liu said, it may as well be the only way and I'm fine with that, considering it'll be well-defined and documented.

> I don't see a way of handling the log using valid output as terminators for entries

Makes sense

**Comment by crassirostris** [comment]
@Random-Liu If we have to escape special symbols anyway, mb it doesn't make sense to introduce a new format, but rather stick with json? Not the most efficient format, but widely-used, extensible and doesn't have problems mentioned in this issue

**Comment by Random-Liu** [comment]
@crassirostris I think initially, we use plain text for performance consideration, and make it easier for the log agent to consume.

However, if we still need to escape eof, I'm not sure whether plain text is still better than json format then.

@yujuhong

**Comment by cyphar** [comment]
If the log format was something like

```
{"time": "2017-05-31T01:13:35+00:00", "stream": "stdout", "data": "this is some stdout\n\nhe"}
{"time": "2017-05-31T01:13:36+00:00", "stream": "stdout", "data": "llo"}
{"time": "2017-05-31T01:13:37+00:00", "stream": "stderr", "data": "ERROR: Appli"}
{"time": "2017-05-31T01:13:38+00:00", "stream": "stdout", "data": "!\n"}
{"time": "2017-05-31T01:13:39+00:00", "stream": "stderr", "data": "cation is dying!\n"}
```

However I think the biggest issue is that this log format doesn't handle the entire `latin1` character set properly (since JSON strings must be valid unicode, you can't give any random `\xXX` code that you want). Maybe some sort of length-prefix format will work better?

**Comment by yujuhong** [comment]
Escaping sounds good to me since we can't come up with an alternative solution.

I am neutral on the json format though. What benefits does it bring?

**Comment by mrunalp** [comment]
cc: @alexlarsson

**Comment by crassirostris** [comment]
@yujuhong Support out of the box from the most popular logging agents. Ability to extend the set of fields is the future, if necessary (e.g. for a custom deployment include additional metadata pieces). Easy to translate from the journald structured logs, could be a middle-ground.

**Comment by yujuhong** [comment]
@crassirostris 

> Support out of the box from the most popular logging agents. 

That's nice to have.

> Ability to extend the set of fields is the future, if necessary (e.g. for a custom deployment include additional metadata pieces). 

I am not sure if we want to pursue that route. Right now the format exists solely to fulfill the requirement of `kubectl logs`, and we need to be careful extending it for any other reason.

> Easy to translate from the journald structured logs, could be a middle-ground.

Could you explain more? 

-----------

> However I think the biggest issue is that this log format doesn't handle the entire latin1 character set properly (since JSON strings must be valid unicode, you can't give any random \xXX code that you want). Maybe some sort of length-prefix format will work better?


@cyphar, what do you mean by the length-prefix format?

**Comment by crassirostris** [comment]
@yujuhong 

> I am not sure if we want to pursue that route. Right now the format exists solely to fulfill the requirement of kubectl logs, and we need to be careful extending it for any other reason.

Not in the core Kuberentes, that's for sure, I agree that it's not a very strong argument, but rather a nice thing to have customizable.

> Could you explain more?

Journald log entry is a set of key-value pairs, so it translates well to json format. Redirecting (e.g. in CRI) journald to json-formatted files will give the following benefits, compared to using raw journald:

- Per-pod log storage management (in the future)
- Working kubectl

And at the same time, the whole set of metadata in the original journald log entry will be available for logging agent to consume. Again, that's not something of a high importance, just an observation.

**Comment by alexlarsson** [comment]
I don't think using json is really helpful here, because we would not be able to persist non-unicode output that way. I also think that escaping things is somewhat painful, because we need to parse and allocate possible extra buffer size when writing the data.

Why don't we just take the existing format:
```
TIME<space>STREAM<space><raw bytes><newline>
```

But, when we get a partial line, we use a slightly different format to mark this. For instance:
```
TIME<space>STREAM;partial<space><raw bytes...><newline>
```

This is efficient to generate, doesn't change existing log formats, and adds support for partial lines, cat:ing binaries and interleaved stderr/stdout writes.

**Comment by Random-Liu** [comment]
@alexlarsson This is easy to generate, we can easily make kubelet to understand it.

I don't know whether it's log agent friendly @crassirostris

**Comment by cyphar** [comment]
@yujuhong 

> @cyphar, what do you mean by the length-prefix format?

The idea would be that every single log entry specifies how many bytes long it is, and that's considered a single "chunk" of the log. Unfortunately it would mean the log isn't really a "nice" NL-delimited log anymore, but the idea would be that each entry looks like

```
<time>\x31<stream name>\x31<number of bytes for this chunk>\x31<raw bytes>\x30
```

[You can pick whatever record separators you like, personally I like the dedicated ASCII characters but w/e.]

**Comment by yujuhong** [comment]
I suggest we add the `PARTIAL` field for now to make the current format usable for 1.9. If in the future we identify clear needs and plans to embed more metadata in the log format, we can add support for json or other format.

The work includes
 - Change kubelet code to parse the PARTIAL format
 - Change the fluentd configuration to ignore the PARTIAL format for now. To reconstruct the lines properly, we would need a plugin. 

Thoughts? @Random-Liu @mrunalp @runcom @vmik

**Comment by crassirostris** [comment]
Sounds fine to me, but I would prefer a field with the line number in the sequence of "chunks", to distinguish between the first chunk and the rest, maybe for better ordering

**Comment by yujuhong** [comment]
@crassirostris would a chunk number help much in this case? The logging agent needs to read sequentially regardless. If the lines could be out of order, the sequence itself would not be sufficient. You would need an ID for each log entry too., e.g., `(ENTRY_ID, CHUNK_SEQ_ID)`.

**Comment by Random-Liu** [comment]
> I suggest we add the PARTIAL field for now to make the current format usable for 1.9. If in the future we identify clear needs and plans to embed more metadata in the log format, we can add support for json or other format.

SGTM.

> Sounds fine to me, but I would prefer a field with the line number in the sequence of "chunks", to distinguish between the first chunk and the rest, and maybe for better ordering on the backend

For example, if log lines have already been messed up for some reason, how do we know `CHUNK 2` belongs to which log entry?
```
2016-10-06T00:17:09.669794202Z stdout CHUNK 1 abcdefg
2016-10-06T00:17:09.669794202Z stdout CHUNK 1 opqrst
2016-10-06T00:17:09.669794203Z stdout CHUNK 2 hijklmn
```

**Comment by crassirostris** [comment]
@yujuhong As you said, logging agent reads logs sequentially, so if needed, id can be generated on the agent side

@Random-Liu In the log file -- to the second, because we assume that in the log file lines are not messed up. When exporting to backend we can use some id to identify chunks that belong to the same log entry

That said, I would prefer for the start just "trim" entries by throwing out everything that's not the first chunk. Distinguishing between the first entry in a row and a full entry would also be useful. E.g. 0 for full entries, 1, 2, etc. for chunked entries

**Comment by Random-Liu** [comment]
@crassirostris My concern is that the length of the chunk number is not fixed. It could grow potentially.

It's a bit more complex to handle non fixed log prefix in the container runtime. And it's possible that the number will grow too long potentially.

If we only need something to identify the first line. We could mark the first line with some special character.

**Comment by Random-Liu** [comment]
Offline discussed with @yujuhong.
@crassirostris Actually, we think we should have a multi-line plugin for CRI purpose (the current fluentd multiline plugin doesn't work for us) before rolling CRI container runtime out to production. If we have that, it doesn't matter whether we could identify the first line or not.

How about this?
* For short term, we add a field to identify whether a line is partial or last line, probably using `P` for `Partial` and `E` for `End` (last line). Fluentd should just match each line now (use `.` for that field).
* Before we roll out CRI container runtime to production, we should have a fluentd plugin ready. If we find it's hard to have such fluentd plugin. We could always define an `S` as the identification of the first line.

**Comment by crassirostris** [comment]
@alexlarsson It's not possible to represent non-unicode lines in JSON using strings, that's true. However, we can use byte arrays as a fallback

@Random-Liu 

> Before we roll out CRI container runtime to production, we should have a fluentd plugin ready. If we find it's hard to have such fluentd plugin. We could always define an S as the identification of the first line.

The desired behavior right now is to set the line limit at 100KB and drop everything except the first line. That's why I wanted to figure out an approach that would allow to distinguish it

**Comment by crassirostris** [comment]
I strongly suggest to prepare some data to based our decisions on. So far I've heard that JSON is ineffective, and that's the main reason not to use it, however, there are no numbers and no comparison with e.g. disk write times. Maybe JSON is 10 times less effective than CRI format, but that compares to even more time consuming writes and then this is just premature optimization

I really hate to break format every now and then, changing all the related components and tests. Let's spend some time once to avoid hitting it again and again in the future

/cc @piosz @fgrzadkowski

**Comment by Random-Liu** [comment]
> The desired behavior right now is to set the line limit at 100KB and drop everything except the first line. That's why I wanted to figure out an approach that would allow to distinguish it

I don't think we should set the limit in container runtime. The fluentd plugin should provide some limit size configuration, we should configure at the fluentd level, instead of container runtime.

**Comment by yujuhong** [comment]
> I strongly suggest to prepare some data to based our decisions on. So far I've heard that JSON is ineffective, and that's the main reason not to use it, however, there are no numbers and no comparison with e.g. disk write times. Maybe JSON is 10 times less effective than CRI format, but that compares to even more time consuming writes and then this is just premature optimization

Yes, I agree we should have get some numbers. 

> I really hate to break format every now and then, changing all the related components and tests. Let's spend some time once to avoid hitting it again and again in the future

I agree that breaking the format is annoying, but I've also not seen "many changes" in the json log format yet (with the exception of the new partial field, which may require more changes in fluentd plugin anyway).


JSON or not, we need to handle the log line size limit and/or multiline reconstruction issue. I think we should separate the two issues in the discussion.

**Comment by Random-Liu** [comment]
Test Data:
```console
$ wc -l cri.log 
8000000 cri.log
$ wc -l json.log 
8000000 json.log
$ tail -n 2 cri.log 
2017-11-17T07:08:25.666705234Z stdout aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
2017-11-17T07:08:25.666705234Z stdout aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
$ tail -n 2 json.log 
{"log":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\n","stream":"stdout","time":"2017-11-17T07:08:25.666705234Z"}
{"log":"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\n","stream":"stdout","time":"2017-11-17T07:08:25.666705234Z"}
$ ls -al *.log
-rw-rw-r-- 1 lantaol lantaol 1352000000 Nov 17 19:58 cri.log
-rw-rw-r-- 1 lantaol lantaol 1608000000 Nov 17 19:56 json.log
```

Test code:
```golang
package main

import "fmt"
import "time"
import "os"
import "io/ioutil"
import "k8s.io/api/core/v1"
import "k8s.io/kubernetes/pkg/kubelet/kuberuntime/logs"

func main() {
	path := os.Args[1]
	fmt.Println("Read log file", path)
	start := time.Now()
	fmt.Println("Start at", start)
	stdout, stderr := ioutil.Discard, ioutil.Discard // os.Stdout, os.Stderr
	options := logs.NewLogOptions(&v1.PodLogOptions{}, time.Now())
	if err := logs.ReadLogs(path, "", options, nil, stdout, stderr); err != nil {
		fmt.Println(err)
	}
	fmt.Println("End at", time.Now())
	fmt.Println("Takes time", time.Since(start))
}
```

Test Result:
```console
$ time ./test cri.log
Read log file cri.log
Start at 2017-11-17 20:03:15.956118258 +0000 UTC m=+0.052589227
End at 2017-11-17 20:03:22.836704164 +0000 UTC m=+6.933175198
Takes time 6.880626239s

real	0m6.938s
user	0m8.380s
sys	0m0.768s

$ time ./test json.log 
Read log file json.log
Start at 2017-11-17 20:03:53.920614689 +0000 UTC m=+0.051326853
End at 2017-11-17 20:04:36.37087276 +0000 UTC m=+42.501584917
Takes time 42.450296196s

real	0m42.506s
user	0m51.100s
sys	0m2.252s
```

RES memory usage for both is around 20MB.

So overall json log format takes 6+ times of time to parse.

**Comment by Random-Liu** [comment]
Writer side:
```console
$ time ./test json.log 
Write log file json.log
Start at 2017-11-17 21:31:49.882973357 +0000 UTC m=+0.071900671
End at 2017-11-17 21:32:32.387818636 +0000 UTC m=+42.576745872
Takes time 42.504879538s

real	0m42.581s
user	0m39.560s
sys	0m14.756s

$ time ./test cri.log 
Write log file cri.log
Start at 2017-11-17 21:34:09.109916721 +0000 UTC m=+0.059430335
End at 2017-11-17 21:34:30.227238731 +0000 UTC m=+21.176752304
Takes time 21.117355593s

real	0m21.182s
user	0m16.596s
sys	0m10.616s
```

RES memory usage for both is around 20MB.

Overall, json log format takes 2+ times of time on the writer side.

Test code:
```
package main

import "fmt"
import "time"
import "os"
import "bytes"
import "io/ioutil"
import "encoding/json"
import "k8s.io/api/core/v1"
import "k8s.io/kubernetes/pkg/kubelet/kuberuntime/logs"

func main() {
	// read()
	writeCRI()
	// writeJSON()
}

func read() {
	path := os.Args[1]
	fmt.Println("Read log file", path)
	start := time.Now()
	fmt.Println("Start at", start)
	stdout, stderr := ioutil.Discard, ioutil.Discard // os.Stdout, os.Stderr
	options := logs.NewLogOptions(&v1.PodLogOptions{}, time.Now())
	if err := logs.ReadLogs(path, "", options, nil, stdout, stderr); err != nil {
		fmt.Println(err)
	}
	fmt.Println("End at", time.Now())
	fmt.Println("Takes time", time.Since(start))
}

func writeCRI() {
	path := os.Args[1]
	fmt.Println("Write log file", path)
	line := []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
	stream := []byte("stdout")
	delimiter := []byte{' '}
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0640)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer f.Close()
	start := time.Now()
	fmt.Println("Start at", start)
	for i := 0; i <= 8000000; i++ {
		timestamp := time.Now().AppendFormat(nil, time.RFC3339Nano)
		data := bytes.Join([][]byte{timestamp, stream, line}, delimiter)
		data = append(data, '\n')
		if _, err := f.Write(data); err != nil {
			fmt.Println("Error:", err)
			return
		}
	}
	fmt.Println("End at", time.Now())
	fmt.Println("Takes time", time.Since(start))
}

type message struct {
	Timestamp time.Time
	Stream    string
	Message   string
}

func writeJSON() {
	path := os.Args[1]
	fmt.Println("Write log file", path)
	line := []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
	stream := "stdout"
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0640)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer f.Close()
	start := time.Now()
	fmt.Println("Start at", start)
	for i := 0; i <= 8000000; i++ {
		m := &message{
			Timestamp: time.Now(),
			Stream:    stream,
			Message:   string(line),
		}
		data, err := json.Marshal(m)
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		data = append(data, '\n')
		if _, err := f.Write(data); err != nil {
			fmt.Println("Error:", err)
			return
		}
	}
	fmt.Println("End at", time.Now())
	fmt.Println("Takes time", time.Since(start))
}
```

**Comment by Random-Liu** [comment]
Overall, if the benchmark method is correct.
JSON log format uses 6+ times CPU on the reader side, and 2+ times CPU on the writer side.
For container runtime and fluentd, the overhead is proportional to container number.

Given so, unless we have strong opinion that we are going to add many metadata in the future, I prefer CRI log format, and I feel like the "sub tag" based extensibility is enough for now.

I mean:
```
TIMESTAMP STREAM TAG CONTENT
2016-10-06T00:17:09.669794202Z stdout P:TAG1:TAG2 log content 1
2016-10-06T00:17:09.669794203Z stdout P:TAG1:TAG2 log content 2
```

Also, as @yujuhong mentioned, to better support long log line (instead of just pick the first line), we need to mark partial lines anyway, no matter we are using json format or current CRI format.

**Comment by feiskyer** [comment]
Could we consider other solutions ( e.g. JSON format, though it is not so efficient) not to break the log format again and again? The current design seems still not expressive, and may be breaked again if more metadata is added in the near future.

**Comment by crassirostris** [comment]
@Random-Liu I run the same code, with or without actual writing to file, results below. I see that writing takes more time that serialization. Ofc, it's a single data point, but it means that faster serialization will not always provide higher logging throughput on the same disk. From my perspective, this makes performance consideration not that important: all extra time spent is on not-time-sensitive paths. From my point view, definitely less important than configurability and not breaking the format for _everyone_ each time it's not expressive enough.

## CRI

### With actual writing

```
$ ./test test.txt
Write log file test.txt
Start at 2017-11-20 14:14:36.659737439 +0100 CET m=+0.000475701
End at 2017-11-20 14:15:17.563096743 +0100 CET m=+40.903834848
Takes time 40.903394073s
```

### Without actual writing

```
$ ./test test.txt
Write log file test.txt
Start at 2017-11-20 14:20:46.907469614 +0100 CET m=+0.000434239
End at 2017-11-20 14:20:53.480465702 +0100 CET m=+6.573430223
Takes time 6.57302608s
```

## JSON

### With actual writing

```
$ ./test test.txt
Write log file test.txt
Start at 2017-11-20 14:16:25.499673498 +0100 CET m=+0.000468196
End at 2017-11-20 14:17:29.235732523 +0100 CET m=+63.736527102
Takes time 1m3.736084818s
```

### Without actual writing

```
$ ./test test.txt
Write log file test.txt
Start at 2017-11-20 14:19:35.605334734 +0100 CET m=+0.000489862
End at 2017-11-20 14:20:01.389440365 +0100 CET m=+25.784595391
Takes time 25.784133618s
```

**Comment by Random-Liu** [comment]
@crassirostris I don't quite understand. We are not talking about the latency/throughput. The benchmark result shows that json format needs more "CPU Time" than cri log format. It will has 6x higher cpu usage.

In our benchmark, because we generate logs so fast that in both cases, they use up 1 core, and json log format uses 6 times more time to finish.

If we don't generate log so fast, it's possible that they could finish at the same time, but json log will probably use 0.6 cores, and cri log format uses 0.1 cores.

I don't know whether we have any latency/through SLO for log. But for kubelet and container runtime, we do care about the cpu/memory usage of the daemons.

**Comment by crassirostris** [comment]
@Random-Liu I understand that resource consumption with JOSN will be higher, that's a payment for the extensibility.

My point is, if using serialization protocol X for logs makes an application block on logging, that's a clear no-go for X. However, in JSON case this is not true: since writing takes more time that serialization, by buffering several lines we're not blocking the application on logging more that writing to disk would.

Then, it's a matter of a trade-off between performance and extensibility :) From my point of view, extensibility is important than somewhat better performance, but I understand arguments for another point too. I'm not sure how to resolve this argument, ask steering committee maybe?

**Comment by Random-Liu** [comment]
@crassirostris We are not saying that we'll not support json log format anytime soon. The whole stack, including `docker`, `kubelet` and `fluentd` will still support json log format.

For new CRI container runtime, we are making the trade off between performance and extensibility. It doesn't seem like extensibility is top priority today, because the requirement is not clear, and the docker json log format is pretty stable overtime. So I think we could go with performance for now.

If in the future, extensibility does become important, and it's clear that we are going to add several more metadata into the log format, we could always switch back to json log format. Everything is still there, we just need to change CRI container runtime to return json log format.

**Comment by crassirostris** [comment]
@Random-Liu 

> It doesn't seem like extensibility is top priority today, because the requirement is not clear, and the docker json log format is pretty stable over time. So I think we should go with performance for now.

I'm not worried about today, I'm more worried about CRI format making it hard/close to impossible to improve extensibility tomorrow. For example, cri-o already implemented the old format, now they have to change it. Next time it may be 10 different runtimes changing format because the current one is not extensible enough. You're trading performance today (arguably on not that important path even) for extensibility tomorrow, which is from my point of view is wrong.

> If we can list a group of metadata we are sure that we are going to add in near future, I think we should choose json log format now. If not, I prefer making minimum change, and keeping better performance for now.

If I could list all metadata pieces that will be required in the future, we wouldn't have the problem with partial lines, right? :)

**Comment by yujuhong** [comment]
Let me try to summarize the discussions. JSON supports better extensibility, but it incurs a higher CPU usage (~6x) compared with the current text format. There are also other methods to address some extensibility issues, e.g., metadata, and log format may not be the place to host the information. We reached a conclusion in 1.9 to continue using the text format, with a new TAG field to address the existing issues, and provide *some* level of extensibility.

The updated log format: 
https://github.com/kubernetes/kubernetes/blob/3a1c9449a956b6026f075fa3134ff92f7d55f812/pkg/kubelet/apis/cri/v1alpha1/runtime/constants.go#L39-L55

We are open to support a different format later when there are clear needs for it. For now, I don't think there is much value keeping the issue open. Closing.

**Comment by StevenACoffman** [comment]
@Random-Liu 	
Is your example still accurate?
```
2016-10-06T00:17:09.669794202Z stdout P:TAG1:TAG2 log content 1
2016-10-06T00:17:09.669794203Z stdout P:TAG1:TAG2 log content 2
```
Is this format documented anywhere?
We are struggling with how to support it in fluent-bit.

The current fluent-bit `crio` parser uses this regex predating the logtag:
```
/^(?<time>.+)\b(?<stream>stdout|stderr)\b(?<log>.*)$/
```
@chlunde Has proposed in fluent/fluent-bit#876 this regex proposal:
```
/^(?<time>.+)\b(?<stream>stdout|stderr)( (?<logtag>[A-Z]))?\b(?<log>.*)$/
```
@kskewes Has this proposed this regex in fluent/fluent-bit#873 :
```
^(?<time>.+) (?<stream>stdout|stderr) (?<partialflag>F|P) (?<message>.*)$
```

If you look at the regex fluentd uses here [Line 128](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/fluentd-elasticsearch/fluentd-es-configmap.yaml#L128) it is currently:
```
/^(?<time>.+) (?<stream>stdout|stderr) [^ ]* (?<log>.*)$/
```
Can you speak to which would best meet your understanding, use case, and experience?

**Comment by Random-Liu** [comment]
@StevenACoffman Sorry, just saw your comment. Here is the document https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-cri-logging.md.

And feel free to ping me on slack if you need me next time. :) Too many github notifications everyday. :P
